### ISSUE TITLE
How to go about integrating mediapipe's hand tracking into an existing Visual Studio C++ project?

### ISSUE THREAD

**Issue by CristianNCC** [google/mediapipe#1162]
I must admit that I am quite a newb when it comes to build systems. So far I've managed to build the hello_word and the hand_tracking examples on Windows (going through quite a few hoops along the way). Of course, building those examples gives me an executable, which I can't integrate in another project.

What I'd like to do is have mediapipe as a dll or something similar so I can pass the framework an image and get joint coordinates in return, ideally. Can someone point me in the right direction?

**Comment by BitLoose** [comment]
An example of this would be useful as I can't see any clear explanation of how mediapipe is integrated into a standard c++ application.

**Comment by ttamas0713** [comment]
> An example of this would be useful as I can't see any clear explanation of how mediapipe is integrated into a standard c++ application.

You can copy or modify the existing file called _demo_run_graph_main.cc_ or _demo_run_graph_main_gpu.cc_ at _/mediapipe/examples/desktop_. Also, _simple_main_graph.cc_ (c++ file of hello_world example) can be useful to learn how to start a graph and get it's output.

**Comment by CristianNCC** [comment]
> 
> 
> > An example of this would be useful as I can't see any clear explanation of how mediapipe is integrated into a standard c++ application.
> 
> You can copy or modify the existing file called _demo_run_graph_main.cc_ or _demo_run_graph_main_gpu.cc_ at _/mediapipe/examples/desktop_. Also, _simple_main_graph.cc_ (c++ file of hello_world example) can be useful to learn how to start a graph and get it's output.

Well, I've tried that in many ways already :/. What I mean is I've tried to changed the Bazelrules for the BUILD file in _/mediapipe/examples/desktop for demo_run_graph_main so it builds a DLL that you can simply plug into another application along with a header file. From what I've seen, the cc_library Bazel rules offer you a "linkstatic" flag that is true by default and it can be set to false. The output binaries didn't seem to differ when I changed that one flag though...

**Comment by jerrysheen** [comment]
I'm also interested in this, have u guys already figure out how to do this?

**Comment by shubhamvad** [comment]
Hi guys, any update on this ? I am also interested to know the solution on this. Thanks in advance.

**Comment by ShravanRai22** [comment]
Even I'm trying to figure out the same, eagerly looking forward to this feature

**Comment by FabianWildgrube** [comment]
Same here. I would be very interested to see how this could be done.

**Comment by vamsee-asu2019** [comment]
Have you guys figured this out ?
I am trying to use the mediapipe  in another c++ project

**Comment by travisjayday** [comment]
Bump. As a fellow noob to the framework, the task of integration into existing C++ projects seems really daunting without guidance.

**Comment by BitLoose** [comment]
I agree, a lot of serious developers will be put off until there is at 
least a working example of how to interface with C==++.

**Comment by sgowroji** [comment]
Hi @CristianNCC, Have you got a chance to go through this [update](https://github.com/google/mediapipe/issues/1839#issuecomment-817253475).

**Comment by carter54** [comment]
Same here. An example of how to integrating mediapipe application into an existing c++ project will be very helpful~

**Comment by BitLoose** [comment]
Any update on a c++ sample for MediaPipe?

MediaPipe is overlooking a lot of professional users and projects for 
not having c++ api examples, how long would it take someone in the MP 
team to do this, half an hour?

Best regards
Rob Colclough
CTO
Sentel Advance Systems - ViewPoint Group
T. US +1 215 240 6030
T. UK +44 (0)20 8144 3150 (Low-rate 03333 442 120)
www.Sentel-Advance.com
Skype: ViewPoint3D
YouTube: www.youtube.com/viewpoint3d
  
Viva3D software offers a fresh and immediate way to create stunning 3D presentations, info TV, content and simulations with real-time data for almost any type of application.

This email and any files transmitted with it are confidential and intended solely for the use of the individual
or entity to whom they are addressed. If you have received this email in error please notify the system manager.
This message contains confidential information and is intended only for the individual named. If you are not the
named addressee you should not disseminate, distribute or copy this e-mail. Please notify the sender immediately
by e-mail if you have received this e-mail by mistake and delete this e-mail from your system. If you are not the
intended recipient you are notified that disclosing, copying, distributing or taking any action in reliance on the
contents of this information is strictly prohibited.

**Comment by hadon** [comment]
Due to limited resources, MediaPipe team is not actively working on this. 

I see some discussion about consuming a bazel build artifacts from cmake at: https://groups.google.com/g/bazel-discuss/c/LP9tyCws_cg?pli=1.  It looks like a custom cmake rule (add_custom_target) can be used to build a bazel target, such as the mediapipe calculator_framework.  One difficulty has been grouping the build artifacts into a single object file.  It looks like bazel team is addressing this difficulty in https://github.com/bazelbuild/bazel/issues/1920.  

One additional difficulty is that, for now, the MediaPipe C++ library only works within a single shared library or DLL, due to global "registry" singletons.  For now, you will need to call the MediaPipe APIs only from within a single DLL, which can in turn export its own public functions.  

The related issue google/mediapipe#1746 provides a hint about building a DLL from bazel, https://stackoverflow.com/questions/49664449/can-bazel-build-dll.

**Comment by BitLoose** [comment]
That's crazy, c++ is the only open standards OO language used in IOT and 
other embedded devices, it is the most flexible and efficient language 
available, reports say its popularity is in the top 3 languages and 
rising, yet MediaPipe doesn't support it.

IMHO you need to assign resources to this issue or your project will be 
handicapped.

Best regards
Rob Colclough
CTO
Sentel Advance Systems - ViewPoint Group
T. US +1 215 240 6030
T. UK +44 (0)20 8144 3150 (Low-rate 03333 442 120)
www.Sentel-Advance.com
Skype: ViewPoint3D
YouTube: www.youtube.com/viewpoint3d
  
Viva3D software offers a fresh and immediate way to create stunning 3D presentations, info TV, content and simulations with real-time data for almost any type of application.

This email and any files transmitted with it are confidential and intended solely for the use of the individual
or entity to whom they are addressed. If you have received this email in error please notify the system manager.
This message contains confidential information and is intended only for the individual named. If you are not the
named addressee you should not disseminate, distribute or copy this e-mail. Please notify the sender immediately
by e-mail if you have received this e-mail by mistake and delete this e-mail from your system. If you are not the
intended recipient you are notified that disclosing, copying, distributing or taking any action in reliance on the
contents of this information is strictly prohibited.

**Comment by willstott101** [comment]
This is still very much a real issue affecting a lot of people. I'm not sure closing it is helpful - even though the mediapipe team doesn't have the resources to approach it right now.

**Comment by BitLoose** [comment]
If anyone interested in doing this work pls email me, ***@***.***

**Comment by MahtabDahaghin** [comment]
any update? I have to use mediapipe in a C++ project, for my master thesis. Any leads would be helpful.

**Comment by FabianWildgrube** [comment]
> any update? I have to use mediapipe in a C++ project, for my master thesis. Any leads would be helpful.

If your project is a standalone and does not need to be integrated into an existing C++ project that uses MSVC (and thus Visual Studio), you could use a dockerized setup to still run it on windows. I did that in this project: https://github.com/FabianWildgrube/PupilSizeTracker. Maybe that helps.

As for integration with a Visual C++ project I've had no luck, though.

**Comment by MahtabDahaghin** [comment]
I have to integrate mediapipe to an existing C++ project in Qt.

**Comment by willstott101** [comment]
You basically need to add a target like this one: https://github.com/willstott101/meta-mediapipe/blob/master/recipes-framework/mediapipe/files/0002-Build-library.patch then copy all the headers from bazel's working directories and try linking to it.

If you're already linking to any of the libs that mediapipe uses you may have trouble - you'll need to make link to those.

FYI this is for a dynamic lib using GPU on linux - I haven't tried on windows you may only need to replace `.so` with `.dll`

**Comment by maletsden** [comment]
Hi, check my PR with an example of creating a FaceMesh Windows DLL example - [Face mesh Windows DLL desktop example](https://github.com/maletsden/mediapipe/pull/1). This example generates standalone .dll that, If further used in my other Visual Studio Project, by simply loading DLL at runtime from C++ (on Windows - [Run-Time Dynamic Linking](https://docs.microsoft.com/en-us/windows/win32/dlls/using-run-time-dynamic-linking))

**Comment by fortminors** [comment]
@maletsden Hello! Thanks for providing that info. How would I build the .dll. I am new to Bazel, and tried to run `bazel build -c opt mediapipe/examples/desktop/face_mesh_dll:face_mesh_lib` but it gave me this output with an error:

```
DEBUG: Rule 'rules_foreign_cc' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "c2cdcf55ffaf49366725639e45dedd449b8c3fe22b54e31625eb80ce3a240f1e"
DEBUG: Repository rules_foreign_cc instantiated at:
  C:/users/fortm/desktop/mediapipe-face_mesh_dll_example/WORKSPACE:42:13: in <toplevel>
Repository rule http_archive defined at:
  C:/users/fortm/_bazel_fortm/hhj5wzn5/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
WARNING: Download from http://mirror.tensorflow.org/github.com/bazelbuild/rules_closure/archive/cf1e44edb908e9616030cc83d085989b8e6cd6df.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET returned 404 Not Found
WARNING: Download from http://mirror.tensorflow.org/github.com/tensorflow/runtime/archive/44c890cd46fd9aad2b3bf4b13c5f36d8cff795c3.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET returned 404 Not Found
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'com_google_absl' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'com_google_protobuf' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'com_google_googletest' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'com_github_gflags_gflags' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'build_bazel_rules_apple' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'build_bazel_rules_swift' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'build_bazel_apple_support' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'bazel_skylib' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/repo.bzl:108:14:
Warning: skipping import of repository 'pybind11' because it already exists.
DEBUG: C:/users/fortm/_bazel_fortm/hhj5wzn5/external/tf_runtime/third_party/cuda/dependencies.bzl:51:10: The following command will download NVIDIA proprietary software. By using the software you agree to comply with the terms of the license agreement that accompanies the software. If you do not agree to the terms of the license agreement, do not use the software.
DEBUG: Rule 'rules_cc' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "86bf1c24e8f859fb5c66d931f33dc8283427e7370d2b0d178658893bcd3c49e9"
DEBUG: Repository rules_cc instantiated at:
  C:/users/fortm/desktop/mediapipe-face_mesh_dll_example/WORKSPACE:36:13: in <toplevel>
Repository rule http_archive defined at:
  C:/users/fortm/_bazel_fortm/hhj5wzn5/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
INFO: Build option --compilation_mode has changed, discarding analysis cache.
INFO: Repository local_execution_config_python instantiated at:
  C:/users/fortm/desktop/mediapipe-face_mesh_dll_example/WORKSPACE:408:14: in <toplevel>
  C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/tensorflow/workspace2.bzl:1087:19: in workspace
  C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/tensorflow/workspace2.bzl:85:27: in _tf_toolchains
  C:/users/fortm/_bazel_fortm/hhj5wzn5/external/tf_toolchains/toolchains/remote_config/configs.bzl:6:28: in initialize_rbe_configs
  C:/users/fortm/_bazel_fortm/hhj5wzn5/external/tf_toolchains/toolchains/remote_config/rbe_config.bzl:158:27: in _tensorflow_local_config
Repository rule local_python_configure defined at:
  C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/py/python_configure.bzl:275:41: in <toplevel>
ERROR: An error occurred during the fetch of repository 'local_execution_config_python':
   Traceback (most recent call last):
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/py/python_configure.bzl", line 209, column 22, in _create_local_python_repository
                _check_python_bin(repository_ctx, python_bin)
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/py/python_configure.bzl", line 143, column 52, in _check_python_bin
                result = raw_exec(repository_ctx, [get_bash_bin(repository_ctx), "-c", cmd])
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/remote_config/common.bzl", line 88, column 26, in get_bash_bin
                bash_bin_path = which(repository_ctx, "bash")
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/remote_config/common.bzl", line 27, column 22, in which
                out = execute(
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/remote_config/common.bzl", line 230, column 13, in execute
                fail(
Error in fail: Repository command failed
INFO: Could not find files for the given pattern(s).
ERROR: Error fetching repository: Traceback (most recent call last):
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/py/python_configure.bzl", line 209, column 22, in _create_local_python_repository
                _check_python_bin(repository_ctx, python_bin)
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/py/python_configure.bzl", line 143, column 52, in _check_python_bin
                result = raw_exec(repository_ctx, [get_bash_bin(repository_ctx), "-c", cmd])
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/remote_config/common.bzl", line 88, column 26, in get_bash_bin
                bash_bin_path = which(repository_ctx, "bash")
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/remote_config/common.bzl", line 27, column 22, in which
                out = execute(
        File "C:/users/fortm/_bazel_fortm/hhj5wzn5/external/org_tensorflow/third_party/remote_config/common.bzl", line 230, column 13, in execute
                fail(
Error in fail: Repository command failed
INFO: Could not find files for the given pattern(s).
ERROR: Analysis of target '//mediapipe/examples/desktop/face_mesh_dll:face_mesh_lib' failed; build aborted: Repository command failed
INFO: Could not find files for the given pattern(s).
INFO: Elapsed time: 0.289s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 11 targets configured)
```

How did you manage to build the DLL? Thanks!

**Comment by fortminors** [comment]
@maletsden Since I've managed to build successfully, I can't find the dll file that it should produce. I was searching for `face_mesh_lib.dll`. I fail to understand where Bazel would put the files.

**Comment by mahtaabdn** [comment]
[https://github.com/google/mediapipe/issues/2586#issue-1005490666](url)

**Comment by BaiL939** [comment]
Thanks for the positive discussion, but mediapipe is not very friendly to C++.But my ability is really insufficient, now the C++ development environment is not built well.But after reading everyone's discussion, I still feel that there is no concrete and practical answer

**Comment by 424778940z** [comment]
Please consider move to CMake or at least provided a way to build full library and headers in Bazel.
The real problem here is Bazel is really limited both on function and documents.

It is the end of 2021, and Bazel as a "build tool" is not able to produce a static library natively is ridiculous.
To make the things worse the official devs says they have NO plan for it at all.

Also for Bazel, there is NO WAY to install related headers to a folder, it may not be an issue for "package based" languages.
But C++ need those headers to use the output library.

And speaking of the document, it is a huge mess.
Some of the pages are ok, some are basiclly empty such as this one https://docs.bazel.build/versions/main/skylark/lib/file_provider.html
```
Extensions > API reference > file_provider
file_provider
An interface for rules that provide files.
```
And the documents are only for basic language wise, there is NO ducment for most of the rules, espicially on how to interact with them.
Even something as common as cc_library, there are document on how to use them.
But there is nothing descipt how to get CCInfo (that contains include and link info) from it, you never find out without dig through source.

**Comment by Jerryzhangzhao** [comment]
You can consider convert the mediapipe .tflite model to onnx or even tensorRT model, and then do the inference with in c++ project.

**Comment by Jerryzhangzhao** [comment]
> @Jerryzhangzhao but for example the face detector model outputs anchors. How do you convert them to bounding rects? I haven't figured it out yet. So there is a post-processing problem

Yes, the post-processing step is important. First of all, from the paper [blazeFace](https://arxiv.org/pdf/1907.05047.pdf) , we know that there are two type 896 anchors:
![image](https://user-images.githubusercontent.com/10524773/137829655-831bb0f5-e4b8-44b3-a424-ea883759e764.png)
with the visualization of the tflite model, the output including classificators [1x896x1] and regressors [1x896x16]
![image](https://user-images.githubusercontent.com/10524773/137829997-22bc771c-f567-4258-a227-2fd92a4ef0e7.png)
With a sigmoid activation applied on classificators would generate the confidence score of each boundingbox, and teh first four element in each 16 output of regressors is [y_offset, x_offset, height, width] of the boudningbox.
The above information can be found in the source code of mediapipe [tflite_tensors_to_detections_calculator.cc](https://github.com/google/mediapipe/blob/master/mediapipe/calculators/tflite/tflite_tensors_to_detections_calculator.cc)
where the code segment：
line 326:
```
// Filter classes by scores.
    for (int i = 0; i < num_boxes_; ++i) {
      int class_id = -1;
      float max_score = -std::numeric_limits<float>::max();
      // Find the top score for box i.
      for (int score_idx = 0; score_idx < num_classes_; ++score_idx) {
        if (ignore_classes_.find(score_idx) == ignore_classes_.end()) {
          auto score = raw_scores[i * num_classes_ + score_idx];
          if (options_.sigmoid_score()) {
            if (options_.has_score_clipping_thresh()) {
              score = score < -options_.score_clipping_thresh()
                          ? -options_.score_clipping_thresh()
                          : score;
              score = score > options_.score_clipping_thresh()
                          ? options_.score_clipping_thresh()
                          : score;
            }
            score = 1.0f / (1.0f + std::exp(-score));
          }
          if (max_score < score) {
            max_score = score;
            class_id = score_idx;
          }
        }
      }
      detection_scores[i] = max_score;
      detection_classes[i] = class_id;
    }
```
line 94:
```
void ConvertAnchorsToRawValues(const std::vector<Anchor>& anchors,
                               int num_boxes, float* raw_anchors) {
  CHECK_EQ(anchors.size(), num_boxes);
  int box = 0;
  for (const auto& anchor : anchors) {
    raw_anchors[box * kNumCoordsPerBox + 0] = anchor.y_center();
    raw_anchors[box * kNumCoordsPerBox + 1] = anchor.x_center();
    raw_anchors[box * kNumCoordsPerBox + 2] = anchor.h();
    raw_anchors[box * kNumCoordsPerBox + 3] = anchor.w();
    ++box;
  }
}
```


line 590：
```
 float y_center = raw_boxes[box_offset];
    float x_center = raw_boxes[box_offset + 1];
    float h = raw_boxes[box_offset + 2];
    float w = raw_boxes[box_offset + 3];
    if (options_.reverse_output_order()) {
      x_center = raw_boxes[box_offset];
      y_center = raw_boxes[box_offset + 1];
      w = raw_boxes[box_offset + 2];
      h = raw_boxes[box_offset + 3];
    }
```

**Comment by fortminors** [comment]
@Jerryzhangzhao Thanks a lot for sharing this! It's very useful

**Comment by asprecic** [comment]
> > You basically need to add a target like this one: https://github.com/willstott101/meta-mediapipe/blob/master/recipes-framework/mediapipe/files/0002-Build-library.patch then copy all the headers from bazel's working directories and try linking to it.
> > If you're already linking to any of the libs that mediapipe uses you may have trouble - you'll need to make mediapipe link to those which is a bit of a fiddle - but doable
> > FYI this is for a dynamic lib using GPU on linux - I haven't tried on windows you may only need to replace `.so` with `.dll`
> 
> HI,
> 
> I have integrated the mediapipe as c++ (hand tracking) qt project for webos (assuming CPU) , however the screen is not rendering with latest frames after polling , it's getting returned
> 
> ```
> size_t frame_timestamp_us = static_cast<double>(cv::getTickCount()) / static_cast<double>(cv::getTickFrequency()) * 1e6;
> 
> absl::Status pstatus = graph.AddPacketToInputStream("input_video", mediapipe::Adopt(input_frame.release())
>                       .At(mediapipe::Timestamp(frame_timestamp_us)));
>                                                    
> // Get the graph result packet, or stop if that fails.
> mediapipe::Packet packet;
> if (!poller.Next(&packet)) return;
> auto& output_frame = packet.Get<mediapipe::ImageFrame>();
> ```
> 
> Could you please help if you faced similar problem on webos or any other platform

I also had that problem where poller would return. It is required that the tflite model is located at "mediapipe/modules/selfie_segmentation/selfie_segmentation.tflite" of wherever your binary is.
This path can be changed if you dig into mediapipe.

**Comment by asprecic** [comment]
> You basically need to add a target like this one: https://github.com/willstott101/meta-mediapipe/blob/master/recipes-framework/mediapipe/files/0002-Build-library.patch then copy all the headers from bazel's working directories and try linking to it.
> 
> If you're already linking to any of the libs that mediapipe uses you may have trouble - you'll need to make mediapipe link to those which is a bit of a fiddle - but doable
> 
> FYI this is for a dynamic lib using GPU on linux - I haven't tried on windows you may only need to replace `.so` with `.dll`

Thank you very much for this help. This was the missing puzzle I needed to produce .so and be able to interface to mediapipe from other projects.

I was able to produce .so on linux, and link to it from another Qt project. I did have to manually copy the header file for interface which I created.

Overall, it is possible to integrate this into existing c++ projects, it's just not easy, simple nor well documented. You're pretty much on your own to make it happen. I had to dig through a lot of mediapipe and make changes that get bazel to build the .so how I want it (stripped of useless unused junk libraries).

**Comment by pkoppise** [comment]
@asprecic  @willstott101 
Now able to run hand-tracking cpu example as an QML application. 
a) Generated libmediapipe.so
b) Resolved all the dependencies (absl) 
c) Created a QML application (backend c++, where i am using mediapipe, opencv) and in .pro file linked all needed dependencies
Thanks a lot for the help

**Comment by asprecic** [comment]
Hey everyone,

I made an example Qt application to demonstrate how to integrate Mediapipe as a dynamic library in Linux.
It runs many examples from Mediapipe examples, including the hand tracking.

You can find the example application here:
https://github.com/asprecic/mediapipe-qt-integration-example

and the corresponding changes in Mediapipe in this fork here:
https://github.com/asprecic/mediapipe

**Comment by liupeiqiHN** [comment]
> Hey everyone,
> 
> I made an example Qt application to demonstrate how to integrate Mediapipe as a dynamic library in Linux. It runs many examples from Mediapipe examples, including the hand tracking.
> 
> You can find the example application here: https://github.com/asprecic/mediapipe-qt-integration-example
> 
> and the corresponding changes in Mediapipe in this fork here: https://github.com/asprecic/mediapipe

I try to build mediapipe in windows, build is successful, but when I run demo, I got an error:
```
[libprotobuf ERROR external/com_google_protobuf/src/google/protobuf/text_format.cc:309] Error parsing text-format mediapipe.CalculatorGraphConfig: 33:100: Could not find type "type.googleapis.com/mediapipe.DetectionsToRenderDataCalculatorOptions" stored in google.protobuf.Any.
F20211205 10:07:29.508597 24436 parse_text_proto.h:32] Check failed: ParseTextProto(input, &result)
*** Check failure stack trace: ***
```

**Comment by problemSolvingProgramming** [comment]
@CristianNCC Excuse me, I'm having your problem. My goal is to be able to use the mediapipe pose tracking program in another C ++ project in Visal Studio 2019. Can you help me how to do this?
https://github.com/google/mediapipe/tree/master/mediapipe/examples/desktop/pose_tracking

Or put the code in a repository.

**Comment by liupeiqiHN** [comment]
I can call mediapipe's python API in my C++ program, but the downside is that I have to integrate a `python embed`, and python API is limited in functionality compared to C++ API.

**Comment by BitLoose** [comment]
I agree, its less than optimal, I'm waiting until I can test this in 
Visual Studio C++ without all the complexities?inefficiency.

contents of this information is strictly prohibited.

**Comment by vishal-ahirwar** [comment]
> @asprecic @willstott101 Now able to run hand-tracking cpu example as an QML application. a) Generated libmediapipe.so b) Resolved all the dependencies (absl) c) Created a QML application (backend c++, where i am using mediapipe, opencv) and in .pro file linked all needed dependencies Thanks a lot for the help
Can you please tell me how you did that??????????////
Your whole procedure like how you build dynamic library ???????///
Thanks in Advanced.

**Comment by Mario-td** [comment]
What I did is building the mediapipe C++ program using bazel, then execute it from a different process and communicate through shared memory.
Here is the demo: https://github.com/Mario-td/Simplified-hand-tracking-with-Mediapipe-CPP

**Comment by problemSolvingProgramming** [comment]
@Mario-td Thank you, please do the same for pose  tracking   .

**Comment by liupeiqiHN** [comment]
> i have successfully compiled a dynamic lib of face landmark detection and used it in visual studio 2019 project.

I can use bazel build a dynamic ".dll", but I don't know how to get a ".lib" to link it. Can you share your experience of using VS2019?

**Comment by BitLoose** [comment]
that's interesting, where can I download? Thxs

**Comment by andrechen** [comment]
i put a visual studio 2019 solution file with few selected example project files [here](https://github.com/andrechen/mediapipe/tree/master/visualstudio). And If you build it (solution), they will run/debug.

all in this fork:
https://github.com/andrechen/mediapipe

**Comment by lilinxiong** [comment]
> i put a visual studio 2019 solution file with few selected example project files [here](https://github.com/andrechen/mediapipe/tree/master/visualstudio). And If you build it (solution), they will run/debug.
> 
> all in this fork: https://github.com/andrechen/mediapipe

That's amazing. I changed some things based on bazel before, so that it can build 32-bit library, so I have a question, can this vs project build 32-bit library?

**Comment by andrechen** [comment]
I don't build 32-bit library or executable. It's prone to have problems when you need to malloc a huge chunk memory from heap. 

If you do need it, you can add new x86 platform in visual studio configuration manager. All dependencies are built from source so they are good, only the pre-built x64 [opencv](https://github.com/andrechen/mediapipe/tree/master/third_party/opencv) library is used. Meaning, you'll need to work on opencv 32-bit library. Good luck.

**Comment by lilinxiong** [comment]
> I don't build 32-bit library or executable. It's prone to have problems when you need to malloc a huge chunk memory from heap.
> 
> If you do need it, you can add new x86 platform in visual studio configuration manager. All dependencies are built from source so they are good, only the pre-built x64 [opencv](https://github.com/andrechen/mediapipe/tree/master/third_party/opencv) library is used. Meaning, you'll need to work on opencv 32-bit library. Good luck.

Yes, I agree, I changed some bazel rules based on the original mediapipe, so that it can build 32-bit library and 64-bit library, and of course the corresponding executable, and then I use it when debugging It's vscode instead of visual studio, which is what I need right now. I still admire you for porting mediapipe into a vs project, it's amazing!

**Comment by andrechen** [comment]
@lilinxiong, Sorry, I was wrong. To make x86 target configs isn't that easy.
I had pushed x86 (32-bit) solution and project files with minimal code changes. All 32-bit samples run but with hiccup. See caveats [here](https://github.com/andrechen/mediapipe/tree/master/visualstudio).

**Comment by andrechen** [comment]
I put all live examples, including **face_detection**, **face_mesh**, **hair_segmentation**, **hand_tracking**, **holistic_tracking**, iris_tracking, object_tracking, pose_tracking and template_matching(?)  into a single executable project [here](https://github.com/andrechen/mediapipe/tree/master/visualstudio/live_demo).

So, you can actually build a single monolithic static mediapipe library to serve your need, but since all calculators, subgraphs and input_streams registry depend on **static** dangling pointers (to mediapipe::RegistrationToken) in the **static** library, by default, those symbols will not be referenced by other object files in the executable (so all registries fail). To workaround this, you can use MSVC linker options /WHOLEARCHIVE, (or --whole-archive option for GCC ld). But, this will drive link (and You) crazy. Zzz...

**Comment by liupeiqiHN** [comment]
@andrechen I see that your project directly include mediapipe's source code, maybe we can repackage mediapipe's functions and compile it as a dynamic/static library.

**Comment by BitLoose** [comment]
In hand_landmarks_to_rect_calculator.cc I would like to check the handedness of the landmarks, but I don't know the tag names to use, and I think there is no documentation.
I've added:
   const auto& alandmarks = cc->Inputs().Tag(kNormalizedLandmarksTag).Get<NormalizedLandmarkList>();
    if (alandmarks.landmark_size() == kNumLandmarks) {
        ...
But how to access the data from the earlier Palm detector which has handedness?

**Comment by rajkundu** [comment]
@asprecic Thank you very much for your work! I was able to get your code working on Windows with a few modifications. Would you be willing to check and see if it still works and integrates with your Qt integration example? Here is the link to my fork: https://github.com/rajkundu/mediapipe

The main modification was just adding `__declspec(dllexport)` to function definitions in `example.h`, but I also tried to make the Bazel build rules a bit "smarter" (detect build on Windows vs. Linux), so I'd love to ensure that they still work on Linux. Thank you again for all your work!

**Comment by asprecic** [comment]
@rajkundu 
Thank you, and nice work yourself.

I am no longer working the same job and not really interested in doing this anymore, so I can't really be bothered to install a dev env on my personal Windows machine.

If you make any PR's, I'm happy to merge them, for community's sake.

**Comment by rajkundu** [comment]
No worries at all! That makes sense. For what it is worth, I did try compiling the `.so` on a Ubuntu machine with a small non-Qt example (reads an input image from disk, runs MP, and writes the MP output image to disk), which worked.

**Comment by rajkundu** [comment]
I ended up creating a generalized C++ shared library for MediaPipe, which I call [LibMP](https://github.com/rajkundu/mediapipe/tree/cpp-lib/mediapipe/examples/desktop/libmp). It exposes MediaPipe's Protobuf-related functions so that, **without Bazel**, a client C++ application can access MediaPipe data types near-natively. That is, while `libexample` runs MP graphs and allows access to their output images, `libmp` allows access to the actual data being rendered by MP on those output images. I've created a very simple example of a LibMP client application here: https://github.com/rajkundu/libmp-example

Similar to `libexample`, `libmp` is itself built as a mediapipe target using Bazel. However, once that is done, `libmp` can be used in any other project alongside Google's Protobuf library - without requiring Bazel. Together, LibMP + Protobuf allow an external application to interact with "raw" MediaPipe data. For example, in `libmp-example`, I run a MediaPipe Face Mesh graph and use its `multi_face_landmarks` output stream to manually draw each of the 478 facial landmarks onto the input image. While I used `mediapipe::NormalizedLandmark` and `mediapipe::NormalizedLandmarkList`, many other MP examples and their data types (e.g., the face detection graph and `mediapipe::Detection`) should work very similarly. Since LibMP is generalized, only the client-side code should need to be adapted, not LibMP itself.

So, hopefully anyone looking to use MediaPipe in their own C++ application can take advantage of this work. While it's perhaps not as ideal as building a static library or full CMake support, this shared library can hopefully serve as a stopgap until those things exist!

**Comment by c1ngular** [comment]
@rajkundu nice job , I have been looking for a way to use mediapipe solutions in golang , but it is too hard , your work showed me hope. in order to make it more portable and accessible ,  a c library preferably with pkg-config support would be truly awesome . Do you think it's possible ?

**Comment by c1ngular** [comment]
@rajkundu yes , a c library would be perfect .

**Comment by rajkundu** [comment]
@c1ngular Unfortunately, with my own work ramping up, I won't have much time to make a C version of LibMP myself. However, I truly think it should not be difficult to do the work yourself and make LibMP into a C library. As I said before, all that is needed is to convert the LibMP class's 7-8 member functions (see `virtual` functions in [LibMP.h](https://github.com/rajkundu/mediapipe/blob/cpp-lib/mediapipe/examples/desktop/libmp/libmp.h)) to static functions. They should accept a `void*` pointer to a LibMP instance, just like the functions in [LibMP.cpp](https://github.com/rajkundu/mediapipe/blob/cpp-lib/mediapipe/examples/desktop/libmp/libmp.cpp) accept a `void*` to a MediaPipe packet and `reinterpret_cast` it. If you do end up taking on this small project, I'm happy to help and answer any questions!

**Comment by c1ngular** [comment]
@rajkundu my C++ knowledge is basically zero  , i am not sure if i can handle this . 
thank your for your work anyway .

**Comment by oUp2Uo** [comment]
@rajkundu Thanks for your great job. And I found it seems that there was memory leak in LibMP. When I run the test in loop for some time, memory usage was continuously increase. Have you tried to process not only one image but image sequneces (e.g. video)? Thanks.

**Comment by rajkundu** [comment]
@oUp2Uo Thank you! And you're completely right - indeed, I came across this issue myself when using LibMP on a video feed. However, the fix is simple. There is a static function `mediapipe::LibMP::DeletePacket` which takes a `void*` to a mediapipe packet and frees its memory. If this freeing is not done, then the packets will rapidly accumulate (each one containing an entire `ImageFrame`) and claim memory. You can also use a smart pointer (see below) to automatically manage this for you.

So, it seems that although I created the `mediapipe::LibMP::DeletePacket` function and pushed it to my mediapipe fork, I forgot to add it to my `libmp-example` repo's code - sorry about that! Until I fix that, here is some example usage from another project of mine:
```cpp
// return landmark XYZ data for all detected faces (or empty vector if no detections)
// dimensions = (# faces) x (# landmarks/face) x 3
// i.e., each landmark is a 3-float array (X,Y,Z), so the middle vector contains 468 or 478 of these
// and the outermost vector is for each detected face in the frame
static std::vector<std::vector<std::array<float, 3>>> get_landmarks(const std::shared_ptr<mediapipe::LibMP>& face_mesh){
    std::vector<std::vector<std::array<float, 3>>> normalized_landmarks;

    // I use a unique_ptr for convenience, so that DeletePacket is called automatically
    // You could also manage deletion yourself, manually:
    // const void* packet = face_mesh->GetOutputPacket("multi_face_landmarks");
    // mediapipe::LibMP::DeletePacket(packet);
    std::unique_ptr<const void, decltype(&mediapipe::LibMP::DeletePacket)> lm_packet_ptr(nullptr, mediapipe::LibMP::DeletePacket);

    // Keep getting packets from queue until empty
    if (face_mesh->GetOutputQueueSize("multi_face_landmarks")){
        while (face_mesh->GetOutputQueueSize("multi_face_landmarks") > 0){
            lm_packet_ptr.reset(face_mesh->GetOutputPacket("multi_face_landmarks"));
        }
    } else {
        return normalized_landmarks; // return empty vector if no packets in queue
    }

    if (lm_packet_ptr.get() == nullptr || mediapipe::LibMP::PacketIsEmpty(lm_packet_ptr.get())){
        return normalized_landmarks; // return empty vector if packets existed but are somehow invalid
    }
    size_t num_faces = mediapipe::LibMP::GetPacketProtoMsgVecSize(lm_packet_ptr.get());

    // Create multi_face_landmarks from packet's protobuf data
    for (int face_num = 0; face_num < num_faces; face_num++){
        // Get reference to protobuf message for face
        const void* lm_list_proto = mediapipe::LibMP::GetPacketProtoMsgAt(lm_packet_ptr.get(), face_num);
        // Get byte size of protobuf message
        size_t lm_list_proto_size = mediapipe::LibMP::GetProtoMsgByteSize(lm_list_proto);

        // Create buffer to hold protobuf message data; copy data to buffer
        std::shared_ptr<uint8_t[]> proto_data(new uint8_t[lm_list_proto_size]);
        mediapipe::LibMP::WriteProtoMsgData(proto_data.get(), lm_list_proto, static_cast<int>(lm_list_proto_size));

        // Initialize a mediapipe::NormalizedLandmarkList object from the buffer
        mediapipe::NormalizedLandmarkList face_landmarks;
        face_landmarks.ParseFromArray(proto_data.get(), static_cast<int>(lm_list_proto_size));

		normalized_landmarks.emplace_back();
		for (const mediapipe::NormalizedLandmark& lm : face_landmarks.landmark()){
			normalized_landmarks[face_num].push_back({lm.x(), lm.y(), lm.z()});
		}
    }

    return normalized_landmarks;
}
```
and for context, here is where/how I called the above function:
```cpp
if(face_mesh->Process(rgb.get(), width, height, mediapipe::ImageFormat::SRGB)){ // Process returns true if successful
    // Wait for processing to complete
    face_mesh->WaitUntilIdle();

    std::vector<std::vector<std::array<float, 3>>> normalized_landmarks = get_landmarks(face_mesh);
    
    // Here you can use normalized_landmarks however you like!
    // Remember everything is normalized to [0, 1], so multiply by input image dimensions as needed
    size_t num_faces = normalized_landmarks.size();
}
```

This code was for the Face Mesh graph and used LibMP to get the X/Y/Z position of each of the 468 (or 478) landmarks for each face in the frame. Sorry it is a bit long, but hopefully the comments help explain every step. If you have any questions, please feel free to ask!

**Comment by oUp2Uo** [comment]
@rajkundu Thanks for your reply. I used the code above to replced the old one. However, memory leak was still existed.

And I found the delete function in the newest libmp-example code. Also I tried to delete the pointer manually, still the same.

https://github.com/rajkundu/libmp-example/blob/main/main.cpp#L123

**Comment by rajkundu** [comment]
Ah yes - thank you for finding the packet deletion in `libmp-example`. I could definitely have documented things better, though. In terms of memory leaks, part of the problem is that I primarily develop LibMP on Windows and have only ever checked simple leaks using valgrind on Linux.

Anyway, 1-2 months ago, I was working with 4K (3840x2160) BGRA images at 30 FPS. and wanted to use LibMP to run the MP face mesh graph. When I first integrated LibMP, my RAM filled up within seconds of beginning to stream video from my cameras. Due to the size and timing of the leak, I knew that un-freed video frames were the problem. So, I spent a few hours diving into LibMP and fixed the issue - or at least I thought. Now, my project processes 4K images at 30 FPS for 2 cameras simultaneously, without issue.

The most important change I made to LibMP was to make `mediapipe::LibMP::Process()` _copy_ the passed image data [(link)](https://github.com/rajkundu/mediapipe/blob/b8cfb25a1f17ca72bf8e106e1d0e233d63460330/mediapipe/examples/desktop/libmp/libmp_impl.cpp#L62-L70):
```cpp
// copy input data to ImgFrame
auto input_frame_for_input = std::make_unique<ImageFrame>();
auto mp_image_format = static_cast<mediapipe::ImageFormat::Format>(image_format);
input_frame_for_input->CopyPixelData(mp_image_format, width, height, data, mediapipe::ImageFrame::kDefaultAlignmentBoundary);

m_frame_timestamp++;

// the created ImageFrame is now owned by the input packet
auto status = m_graph.AddPacketToInputStream(m_input_stream, mediapipe::Adopt(input_frame_for_input.release()).At(mediapipe::Timestamp(m_frame_timestamp)));
```

With this change, my understanding of the way `Process()` now works is this:
- A new `ImageFrame` is created, and the input image data is copied to its buffer.
- An input packet to the MP graph is constructed to feed the `ImageFrame` into the graph. During construction, `mediapipe::Adopt` gives ownership of the `ImageFrame` to this input packet.
- Since the input packet owns the `ImageFrame`, whenever the input packet is destroyed, the `ImageFrame` and its underlying buffer will be freed.

All of this should mean that as soon as `Process()` returns, it is safe to delete the original image data. In my own code, `rgb` is a dynamically-allocated array managed by a smart pointer: `auto rgb = std::shared_ptr<uint8_t[]>(new uint8_t[height * width * 3]);`. So, within a few lines of `Process()` returning, this smart pointer goes out of scope, and the original input image to the MP graph is freed. The only thing left is to get output packets from the MP graph using the code above and then free their data using `mediapipe::LibMP::Process()`.

My understanding of MediaPipe is rather limited, so please let me know if you see any errors in what I've said. Again, could you provide more details on the size and nature of the memory leak you are experiencing?

**Comment by oUp2Uo** [comment]
Hi @rajkundu , thanks for your replay and happy new year.

I am also developing on Windows. And I just change your example code to process video and show result with OpenCV.
I have seen the delete fucntion, and I thought it should work.
But still my code run memory increase fast. (2k video several seconds later memory will be full)

I attached the code in zip file.

[libmpopencvtest.zip](https://github.com/google/mediapipe/files/10328235/libmpopencvtest.zip)

**Comment by rajkundu** [comment]
@oUp2Uo I think I figured it out. LibMP actually seems to be working as intended. Instead, I think the issue is that the [`face_mesh_desktop_live`](https://github.com/google/mediapipe/blob/9252a025e5604cb61b11cbf23943dc7fb9e6f679/mediapipe/graphs/face_mesh/face_mesh_desktop_live.pbtxt) graph - which `libmp-example` and your code both use - renders output video frames of its own via its `FaceRendererCpu` subgraph. I think these output frames are the cause of the memory leak, which can be fixed in a few ways:

1. If you are not using this output video stream at all, then don't enable it when setting up the graph (and/or delete it from the graph's protobuf definition entirely):
    
    ```cpp
    // face_mesh->AddOutputStream("output_video"); // MP output render stream - DISABLED
    face_mesh->AddOutputStream("multi_face_landmarks"); // MP landmark XYZ stream
    face_mesh->Start();
    ```

2. You can leave the output video stream enabled and flush it manually, e.g.:

    ```cpp
    // flush the MP graph's "output_video" output stream
    std::unique_ptr<const void, decltype(&mediapipe::LibMP::DeletePacket)> out_vid_frame_packet_ptr(nullptr, mediapipe::LibMP::DeletePacket);
    // Keep getting packets from queue until empty
    if (face_mesh->GetOutputQueueSize("output_video")) {
        while (face_mesh->GetOutputQueueSize("output_video") > 0) {
            out_vid_frame_packet_ptr.reset(face_mesh->GetOutputPacket("output_video"));
        }
    }
    out_vid_frame_packet_ptr.reset();
    ```

I tested both options on my own system and they worked, with memory usage stabilizing around 70 MB instead of slowly increasing to 1 GB and beyond.

**Comment by rajkundu** [comment]
@oUp2Uo Inspired by our discussion, I've updated [`libmp-example`](https://github.com/rajkundu/libmp-example) to process video frames instead of reading/writing static images from disk. I originally chose to process images for the sake of simplicity, but I now realize that this leap from still images to video isn't as trivial as I thought. With this change, I think that `libmp-example` is now a much better starting point for those looking to use MediaPipe without Bazel and/or with CMake.

**Comment by oUp2Uo** [comment]
@rajkundu I tried to remove the line of video output stream (`face_mesh->AddOutputStream("output_video");`), and the memory usage was not increasing after that. I will do some long time test to confirm this. Thank you very much!

**Comment by rajkundu** [comment]
No problem. Thank you as well! :)

**Comment by oUp2Uo** [comment]
@rajkundu And there is another small bug in the header file:
https://github.com/rajkundu/mediapipe/blob/cpp-lib/mediapipe/examples/desktop/libmp/libmp.h#L90
`LibMP::` is not needed here,
anyway, this will not cause error in VS :)

**Comment by rajkundu** [comment]
Good find - thank you! I must have copied over the function signature from the source file and forgotten to delete that. Will fix it now!

**Comment by jinfagang** [comment]
@rajkundu Hello, nice work on port a lib of mediapipe without bazel. But how did u get these header files out? I want using mediapipe in my proejct, but I don't want copy all bazel0-bin folder to anyother machine I need to build this, is there a way to make a clean header file folder which can be used as include directory?

**Comment by jinfagang** [comment]
@rajkundu Can u tell me which part did u changed? Your fork now conflicts with master branch, you should merge googles master for users better to use, and not conflict with google's new model updated.

**Comment by rajkundu** [comment]
> @rajkundu Hello, nice work on port a lib of mediapipe without bazel. But how did u get these header files out? I want using mediapipe in my proejct, but I don't want copy all bazel0-bin folder to anyother machine I need to build this, is there a way to make a clean header file folder which can be used as include directory?

Unfortunately, my current solution is the easiest one I've found. If I could make MediaPipe into a header-only library, then trust me, I would...

> @rajkundu Can u tell me which part did u changed? Your fork now conflicts with master branch, you should merge googles master for users better to use, and not conflict with google's new model updated.

Check out the [commit history](https://github.com/rajkundu/mediapipe/commits)? That will tell you every single change I've made, and in more detail than I could ever explain. Also, once you clone my fork, you should be able to merge google's master branch into the repository yourself - e.g., by running `git fetch google` then `git merge google/master`. However, I agree that it would be ideal for my repository's upstream branch to be google/mediapipe. I've contacted GitHub support for help on this (my understanding is that I need them to "reroute" my fork).

**Comment by rajkundu** [comment]
I'm happy to report that my fork is now a direct fork of google/mediapipe, which should simplify many things. Now, my fork's changes relative to google/mediapipe can be seen easily, [directly on GitHub](https://github.com/google/mediapipe/compare/master...rajkundu:mediapipe:master). @jinfagang, hopefully, this answers your second question a bit better! As for the first question - if anyone has ideas/suggestions/etc. for better integrating MediaPipe without Bazel, please share! :)

**Comment by jinfagang** [comment]
@rajkundu Hello, I have based on your fork modified a version.
Now it works like a charm, single header file only, I added some API easily convert data from Packet without many proto write.

Now, I have got another question.

I try using holistic graph, the face mesh normal, but the pose didn't get update (seems stucked at some frame)

![image](https://user-images.githubusercontent.com/21303438/216216806-9193e334-2d97-4896-ae42-faebe5aacdf9.png)

Do u know why?

**Comment by rajkundu** [comment]
> @rajkundu Hello, I have based on your fork modified a version. Now it works like a charm, single header file only, I added some API easily convert data from Packet without many proto write.

I'm a C++ beginner, so my understanding of C++ is rather limited. However, when designing LibMP, there were some tradeoffs that I had to consider; I often prioritized compatibility when making design choices. One thing I learned early-on is that [passing C++ objects across DLL boundaries is undefined](https://stackoverflow.com/questions/22797418/how-do-i-safely-pass-objects-especially-stl-objects-to-and-from-a-dll). Apparently, this is due to C++'s lack of a standardized Application Binary Interface (ABI). To get around this, all of the functions in [libmp.h](https://github.com/rajkundu/mediapipe/blob/master/mediapipe/examples/desktop/libmp/libmp.h) return simple, C data types.

Regarding the following functions which you added to your fork of LibMP:
```cpp
virtual const mediapipe::NormalizedLandmarkList DirectlyGetNormalizedLandmarkList(const char *outputStream) = 0;
virtual const std::vector<mediapipe::NormalizedLandmarkList> DirectlyGetNormalizedLandmarkListVec(const char *outputStream) = 0;
virtual cv::Mat *DirectlyGetAsCvMatWrong(const char *outputStream) = 0;
virtual cv::Mat *DirectlyGetAsCvMat(const char *outputStream) = 0;
virtual cv::Mat DirectlyGetAsCvMat2(const char *outputStream) = 0;
```
Due to passing complex C++ data types/objects, I believe that this code may be unstable. It only works if you use compatible (i.e., the same) compilers to build LibMP and your LibMP "client" application. It seems that this worked for your particular project/build environment/machine, but my understanding is that this is a coincidence - that others may not have the same luck as you. So, while it is certainly not ideal to be casting to and reinterpreting from `void*`, etc., LibMP does this to ensure that it works across as many different build environments as possible. As long as you're on the same platform, you should be able to compile LibMP with one compiler (i.e., not have to care what Bazel uses), then build your client application with any compiler of your choice.

My overall goal with LibMP was pretty much to interact with Bazel as little as possible - a "compile once, run anywhere" sort-of approach. Because it is a shared library, you should only have to compile LibMP once using Bazel. Then, you should be able to import it into any number of other projects on that same machine. Not only that, but if I understand correctly, then LibMP DLLs/SOs can actually be distributed (along with header files), so that others don't have to build LibMP themselves. In theory, after building LibMP, I should be able to compress all files (resolving Bazel symlinks) into an archive, which others using the same platform can simply unzip for use in their own projects. (If anyone is interested in this, please let me know!)

So, in the interest of maximizing compatibility, LibMP is limited to C data types and is therefore somewhat inconvenient to use. Of course, I encourage you to make improvements/optimizations for your own use case of LibMP, but do be aware of some potential limitations of shared libraries, especially those mentioned here.

PRs are absolutely _more than welcome_ to LibMP, but I don't think they should come at the cost of compatibility.

**Comment by jinfagang** [comment]
Hello, thank u for answer. I have solved all problems, now it works like a charm

**Comment by rajkundu** [comment]
@jinfagang Yay - glad to hear that! 😁

**Comment by jinfagang** [comment]
@rajkundu Hello, for god sake I hit another problem now..

I don't why mediapipe didn't support windows11, but it actually have problems on link:


<img width="1132" alt="image" src="https://user-images.githubusercontent.com/21303438/216755260-2fba0356-eb7b-4df7-84ba-c6a72b48005f.png">


Do u got a chance test on windows11? Can u built it win11?

As many many users now using win11, it should be a MUST step to make this software forward now.

**Comment by rajkundu** [comment]
I don’t think it’s a Windows 11 problem, because I primarily developed LibMP on a Windows 11 machine (Intel x64) myself. In terms of the compiler, I used MSVC 17.3.5.

Which compiler are you using? Are you setting Bazel environmental variables properly? See the `build_libmp_win` scripts in the top level of my MediaPipe fork, and see [this page](https://bazel.build/configure/windows) for more information on using Bazel with Windows.

**Comment by rajkundu** [comment]
I think we should move discussion to rajkundu/libmp-example#2 to avoid (further) polluting this thread. Sorry, everyone!

**Comment by jinfagang** [comment]
@rajkundu  I maybe because of visual studio version, am using 2019 is fine, 2022 got above errors.

**Comment by oUp2Uo** [comment]
Hello @rajkundu , thanks for your great work and I'm using LibMP well now.
I have one more question:
Now we can get **multiFaceLandmarks** from the result by libmp.h, in which we can get the landmarks.
Could we also get **multiFaceGeometry** from the result? In that case, we can get the rotaion matrix of the faces, I guess?
I thought **multiFaceGeometry** was similar to **multiFaceLandmarks**.
I have tried to read and understand the code, but still do now know how to add it.
Have you have any idea about this? Thanks.

**Comment by rajkundu** [comment]
@oUp2Uo In general, to access any MediaPipe output stream such as multi_face_geometry, you must first ensure that it is defined in your MediaPipe graph. E.g., in my libmp-example repository, I am using a slightly modified version of MediaPipe's [face_mesh_desktop_live](face_mesh_desktop_live.pbtxt) example graph, so only the multi_face_landmarks output stream is currently defined.

So, my first step to incorporate multi_face_geometry would be to find a MediaPipe example graph(s) that uses it - e.g., [single_face_geometry_from_landmarks_gpu.pbtxt](https://github.com/google/mediapipe/blob/master/mediapipe/graphs/face_effect/subgraphs/single_face_geometry_from_landmarks_gpu.pbtxt). Now, try to manually combine the necessary parts of these graphs' configs with the face_mesh_desktop_live graph which you're currently using for landmarks. Copy over nodes and adjust their connections as needed. For example, the `ImagePropertiesCalculator` in `single_face_geometry_from_landmarks_gpu` takes in a GPU image, but try passing it the `throttled_input_video` stream instead.

Here is my first draft of a combined graph:
```
# MediaPipe graph that performs face mesh with TensorFlow Lite on CPU.
# Input image. (ImageFrame)
input_stream: "input_video"
# Output image with rendered results. (ImageFrame)
output_stream: "output_video"
# Collection of detected/processed faces, each represented as a list of
# landmarks. (std::vector<NormalizedLandmarkList>)
output_stream: "multi_face_landmarks"

### Experimental face geometry stuff ###
# A list of geometry data for a single detected face.
#
# NOTE: there will not be an output packet in this stream for this particular
# timestamp if none of faces detected.
#
# (std::vector<face_geometry::FaceGeometry>)
output_stream: "multi_face_geometry"

# Throttles the images flowing downstream for flow control. It passes through
# the very first incoming image unaltered, and waits for downstream nodes
# (calculators and subgraphs) in the graph to finish their tasks before it
# passes through another image. All images that come in while waiting are
# dropped, limiting the number of in-flight images in most part of the graph to
# 1. This prevents the downstream nodes from queuing up incoming images and data
# excessively, which leads to increased latency and memory usage, unwanted in
# real-time mobile applications. It also eliminates unnecessarily computation,
# e.g., the output produced by a node may get dropped downstream if the
# subsequent nodes are still busy processing previous inputs.
node {
	calculator: "FlowLimiterCalculator"
	input_stream: "input_video"
	input_stream: "FINISHED:output_video"
	input_stream_info: {
		tag_index: "FINISHED"
		back_edge: true
	}
	output_stream: "throttled_input_video"
}
# Defines side packets for further use in the graph.
node {
	calculator: "ConstantSidePacketCalculator"
	output_side_packet: "PACKET:0:num_faces"
	output_side_packet: "PACKET:1:use_prev_landmarks"
	output_side_packet: "PACKET:2:with_attention"
	node_options: {
		[type.googleapis.com/mediapipe.ConstantSidePacketCalculatorOptions]: {
			packet { int_value: 1 }
			packet { bool_value: true }
			packet { bool_value: true }
		}
	}
}
# Subgraph that detects faces and corresponding landmarks.
node {
	calculator: "FaceLandmarkFrontCpu"
	input_stream: "IMAGE:throttled_input_video"
	input_side_packet: "NUM_FACES:num_faces"
	input_side_packet: "USE_PREV_LANDMARKS:use_prev_landmarks"
	input_side_packet: "WITH_ATTENTION:with_attention"
	output_stream: "LANDMARKS:multi_face_landmarks"
	output_stream: "ROIS_FROM_LANDMARKS:face_rects_from_landmarks"
	output_stream: "DETECTIONS:face_detections"
	output_stream: "ROIS_FROM_DETECTIONS:face_rects_from_detections"
}
# Subgraph that renders face-landmark annotation onto the input image.
node {
	calculator: "FaceRendererCpu"
	input_stream: "IMAGE:throttled_input_video"
	input_stream: "LANDMARKS:multi_face_landmarks"
	input_stream: "NORM_RECTS:face_rects_from_landmarks"
	input_stream: "DETECTIONS:face_detections"
	output_stream: "IMAGE:output_video"
}

### Experimental face geometry stuff ###
### Adapted from https://github.com/google/mediapipe/blob/master/mediapipe/graphs/face_effect/subgraphs/single_face_geometry_from_landmarks_gpu.pbtxt

# Generates an environment that describes the current virtual scene.
node {
	calculator: "FaceGeometryEnvGeneratorCalculator"
	output_side_packet: "ENVIRONMENT:environment"
	node_options: {
		[type.googleapis.com/mediapipe.FaceGeometryEnvGeneratorCalculatorOptions] {
			environment: {
				origin_point_location: TOP_LEFT_CORNER
				perspective_camera: {
					vertical_fov_degrees: 63.0	# 63 degrees
					near: 1.0	# 1cm
					far: 10000.0	# 100m
				}
			}
		}
	}
}

# Extracts a single set of face landmarks associated with the most prominent
# face detected from a collection.
node {
	calculator: "SplitNormalizedLandmarkListVectorCalculator"
	input_stream: "multi_face_landmarks"
	output_stream: "face_landmarks"
	node_options: {
		[type.googleapis.com/mediapipe.SplitVectorCalculatorOptions] {
			ranges: { begin: 0 end: 1 }
			element_only: true
		}
	}
}
# Extracts the input image frame dimensions as a separate packet.
node {
	calculator: "ImagePropertiesCalculator"
	input_stream: "IMAGE:throttled_input_video"
	output_stream: "SIZE:input_image_size"
}
# Applies smoothing to the single set of face landmarks.
node {
	calculator: "FaceLandmarksSmoothing"
	input_stream: "NORM_LANDMARKS:face_landmarks"
	input_stream: "IMAGE_SIZE:input_image_size"
	output_stream: "NORM_FILTERED_LANDMARKS:smoothed_face_landmarks"
}
# Puts the single set of smoothed landmarks back into a collection to simplify
# passing the result into the `FaceGeometryFromLandmarks` subgraph.
node {
	calculator: "ConcatenateNormalizedLandmarkListVectorCalculator"
	input_stream: "smoothed_face_landmarks"
	output_stream: "multi_smoothed_face_landmarks"
}
# Computes face geometry from face landmarks for a single face.
node {
	calculator: "FaceGeometryFromLandmarks"
	input_stream: "MULTI_FACE_LANDMARKS:multi_smoothed_face_landmarks"
	input_stream: "IMAGE_SIZE:input_image_size"
	input_side_packet: "ENVIRONMENT:environment"
	output_stream: "MULTI_FACE_GEOMETRY:multi_face_geometry"
}
```

I haven't tested it at all, but hopefully it can be a starting point for you. Let me know if I can clarify any further!

**Comment by oUp2Uo** [comment]
@rajkundu Thanks for your infomation! I will try to understand this.

**Comment by oUp2Uo** [comment]
@rajkundu I have a question: In graph, landmark results is **output_stream: "LANDMARKS:multi_face_landmarks"**, how do you know that should use **mediapipe::NormalizedLandmarkList** to get the data?
I am asking this because I tried but I do not know which class should I use to get the data of **output_stream: "MULTI_FACE_GEOMETRY:multi_face_geometry"**.

**Comment by jinfagang** [comment]
I just got one single question, how to specific mediapipe model path root?

**Comment by rajkundu** [comment]
@oUp2Uo Thank you very much for sharing everything! Thanks to your helpful comment and edits, I have now added face geometry to LibMP. Hopefully others can now use it easily as well! :)

**Comment by Chnoblouch** [comment]
We integrated MediaPipe into our CMake project by creating a [C wrapper](https://github.com/cpvrlab/libmediapipe) for it. It works the same way as rajkundu's [LibMP](https://github.com/rajkundu/mediapipe.git), by building a library with Bazel, which can then be used in CMake/Visual Studio/Xcode projects. The library consists of a shared object (```libmediapipe.so```) and a single header (```mediapipe.h```). It currently runs on Linux, Windows and MacOS.

This might be useful for people who want to use MediaPipe in languages that are currently not supported.

**Comment by rajkundu** [comment]
@Chnoblouch That is awesome! I know at least @c1ngular [was looking for something like this](https://github.com/google/mediapipe/issues/1162#issuecomment-1319626539), and I'm sure there are others. Thank you for sharing!

**Comment by c1ngular** [comment]
@rajkundu @Chnoblouch  well done  , thank you both .

**Comment by Daumas-hugo** [comment]
With the end of the support of legacy solutions, are these solutions still a good idea to use for long term app development ?
And if we need a lot of optimization, is it still a good idea to go after mediapipe c++ integration ?

**Comment by kuaashish** [comment]
@CristianNCC,

We are no longer able to provide such configuration from our end. However, We should ship pre-built binaries for Windows if we want better support. Thank you
